plotly/dashboard_objs/dashboard_objs.py

```diff
@@ -0,0 +1,391 @@
+                try:
+                    max_id = max(self.box_ids_to_path.keys())
+                except ValueError:
+                    max_id = 0
+                self.box_ids_to_path[max_id + 1] = list(node[1])
+
+    def _insert(self, box_or_container, path):
+        """Performs user-unfriendly box and container manipulations."""
+        if any(first_second not in ['first', 'second'] for first_second in path):
+            raise exceptions.PlotlyError(
+                "Invalid path. Your 'path' list must only contain "
+                "the strings 'first' and 'second'."
+            )
+
+        if 'first' in self['layout']:
+            loc_in_dashboard = self['layout']
+            for index, first_second in enumerate(path):
+                if index != len(path) - 1:
+                    loc_in_dashboard = loc_in_dashboard[first_second]
+                else:
+                    loc_in_dashboard[first_second] = box_or_container
+
+        else:
+            self['layout'] = box_or_container
+
+    def get_box(self, box_id):
+        """Returns box from box_id number."""
+        self._assign_boxes_to_ids()
+
+        loc_in_dashboard = self['layout']
+        for first_second in self.box_ids_to_path[box_id]:
+            loc_in_dashboard = loc_in_dashboard[first_second]
+        return loc_in_dashboard
+
+    def _path_to_box(self, path):
+        """Returns box from specified path."""
+        self._assign_boxes_to_ids()
+
+        loc_in_dashboard = self['layout']
+        for first_second in path:
+            loc_in_dashboard = loc_in_dashboard[first_second]
+        return loc_in_dashboard
+
+    def get_preview(self):
+        """Returns JSON and HTML respresentation of the dashboard."""
+        # assign box_ids
+        self._assign_boxes_to_ids()
+
+        # print JSON
+        pprint.pprint(self)
+
+        # construct HTML dashboard
+        x = 0
+        y = 0
+        box_w = master_width
+        box_h = master_height
+        html_figure = copy.deepcopy(dashboard_html)
+        path_to_box_specs = {}  # used to store info about box dimensions
+        # add first path
+        first_box_specs = {
+            'top_left_x': x,
+            'top_left_y': y,
+            'box_w': box_w,
+            'box_h': box_h
+        }
+        path_to_box_specs[tuple(['first'])] = first_box_specs
+
+        # generate all paths
+        all_nodes = []
+        node_gen = node_generator(self['layout'])
+
+        finished_iteration = False
+        while not finished_iteration:
+            try:
+                all_nodes.append(node_gen.next())
+            except StopIteration:
+                finished_iteration = True
+
+        all_paths = []
+        for node in all_nodes:
+            all_paths.append(list(node[1]))
+        if ['second'] in all_paths:
+            all_paths.remove(['second'])
+
+        max_path_len = max(len(path) for path in all_paths)
+        # search all paths of the same length
```

Comment:
Why did I write that?

I guess I don't need that comment. I could get rid of it, as I'm just going through all the paths and iteratively picking the paths with a given length (length 1, length 2, ...)